src/core/GameStateManager.ts

```diff
@@ -1,41 +1,82 @@
-        this.board = board;
+    constructor(moveGenerator: MoveGenerator) {
+        this.moveGenerator = moveGenerator;
         this.state = GameState.Default;
     }
 
-    getBoardState(king: King, enemyMoves: Move[]): GameState {
-        this.updateState(king, enemyMoves);
+    getBoardState(board: Board, king: King): GameState {
+        this.updateState(board, king);
         return this.state;
     }
 
-    private updateState(king: King, enemyMoves: Move[]): void {
+    private updateState(board: Board, king: King): void {
+        /* Color of the enemy of the considered king */
+        const enemyColor: Color = king.color === Color.White? Color.Black : Color.White;
+
+        const enemyMoves: Move[] = this.moveGenerator.generateForColor(enemyColor);
+        const allowedMoves: Move[] = this.moveGenerator.generateForColor(king.color);
+
         if(this.isChecked(king, enemyMoves)) {
             this.state = GameState.Check;
+
+            /* If we have inability to move, then its a checkmate! */
+            if(!this.isAbleToMove(board, king, allowedMoves)) {
+                this.state = GameState.Checkmate;
+            }
+        } else if(!this.isAbleToMove(board, king, allowedMoves)) {
+            /* Unable to move but without check -> stalemate ! */
+            this.state = GameState.Stalemate;
+        } else {
+            this.state = GameState.Default;
         }
     }
 
     private isChecked(king: King, enemyMoves: Move[]): boolean {
-        console.log(enemyMoves);
-
         for(const enemyMove of enemyMoves) {
-            console.log(enemyMove.to);
-            console.log(king.position);
             if(enemyMove.to.x === king.position.x && enemyMove.to.y === king.position.y && enemyMove.isCapture) {
                 return true;
             }
         }
 
         return false;
     }
+
+    private isAbleToMove(board: Board, king: King, allowedMoves: Move[]): boolean {
+        /* Color of the enemy of the considered king */
+        const enemyColor: Color = king.color === Color.White? Color.Black : Color.White;
+
+        /* Simulate every allowed move to see if exists legal move */
+        for(const move of allowedMoves) {
+            /* For puprose of move simulating we take deep copy */
+            const tempBoard = _.cloneDeep(board);
+            /* Simulate move */
+            tempBoard.movePiece(move.from, move.to);
+            
+            /* But since generator takes reference to board, its rational to have temporary generator for simulation */
+            const tempGenerator: MoveGenerator = new MoveGenerator(tempBoard);
+            /* Take enemy moves after simulated move */
+            const enemyMoves: Move[] = tempGenerator.generateForColor(enemyColor);
+
+            /* If there is no longer check after simulated move then this move is helping to espace from a checkmate */
+            if(!this.isChecked(tempBoard.findKing(king.color), enemyMoves)) {
```

Comment:
I thnik we can reuse king that we have here from function parameters - we dont have to look for it again